Sources/ArgumentParser/Parsable Types/ParsableArgumentsValidation.swift

```diff
@@ -13,13 +13,99 @@ fileprivate protocol ParsableArgumentsValidator {
+      do {
+        try validator.validate(self)
+        return nil
+      } catch {
+        return error
+      }
+    }
+    if errors.count > 0 {
+      throw ParsableArgumentsValidationError(parsableArgumentsType: self, underlayingErrors: errors)
+    }
+  }
+}
+
+fileprivate extension ArgumentSet {
+  var firstPositionalArgument: ArgumentDefinition? {
+    switch content {
+    case .arguments(let arguments):
+      return arguments.first(where: { $0.isPositional })
+    case .sets(let sets):
+      return sets.first(where: { $0.firstPositionalArgument != nil })?.firstPositionalArgument
+    }
+  }
+  
+  var firstRepeatedPositionalArgument: ArgumentDefinition? {
+    switch content {
+    case .arguments(let arguments):
+      return arguments.first(where: { $0.isRepeatingPositional })
+    case .sets(let sets):
+      return sets.first(where: { $0.firstRepeatedPositionalArgument != nil })?.firstRepeatedPositionalArgument
+    }
+  }
+}
+
+/// For positional arguments to be valid, there must be at most one
+/// positional array argument, and it must be the last positional argument
+/// in the argument list. Any other configuration leads to ambiguity in
+/// parsing the arguments.
+struct PositionalArgumentsValidator: ParsableArgumentsValidator {
+  
+  struct Error: Swift.Error, CustomStringConvertible {
+    let repeatedPositionalArgument: String
+    let positionalArgumentFollowingRepeated: String
+    var description: String {
+      "Can't have a positional argument `\(positionalArgumentFollowingRepeated)` following an array of positional arguments `\(repeatedPositionalArgument)`."
+    }
+  }
+  
+  static func validate(_ type: ParsableArguments.Type) throws {
+    let sets: [ArgumentSet] = Mirror(reflecting: type.init())
+      .children
+      .compactMap { child in
+        guard
+          var codingKey = child.label,
+          let parsed = child.value as? ArgumentSetProvider
+          else { return nil }
+        
+        // Property wrappers have underscore-prefixed names
+        codingKey = String(codingKey.first == "_" ? codingKey.dropFirst(1) : codingKey.dropFirst(0))
+        
+        let key = InputKey(rawValue: codingKey)
+        return parsed.argumentSet(for: key)
+    }
+    
+    guard let repeatedPositional = sets.firstIndex(where: { $0.firstRepeatedPositionalArgument != nil })
+      else { return }
+    let positionalFollowingRepeated = sets[repeatedPositional...]
+      .dropFirst()
+      .first(where: { $0.firstPositionalArgument != nil })
+    
+    if let positionalFollowingRepeated = positionalFollowingRepeated {
+      let firstRepeatedPositionalArgument: ArgumentDefinition = sets[repeatedPositional].firstRepeatedPositionalArgument!
+      let positionalFollowingRepeatedArgument: ArgumentDefinition = positionalFollowingRepeated.firstPositionalArgument!
+      throw Error(repeatedPositionalArgument: firstRepeatedPositionalArgument.valueName,
```

Comment:
What do you think about using `valueName` here.
An argument `numberOfItems` will be reported as `number-of-items` using `valueName`. Should we use `help.keys.first?.rawValue`?

Reply:
Good call — the value name is user-facing; we want to use the actual property names declared in the type here if possible.